### PR TITLE
✨ Source Okta: Authentication with private key (OAuth 2.0) support

### PR DESCRIPTION
**Pull request by evyros**
## What
Adding support of authenticating with Okta using a private key (OAuth2.0) - without affecting current supported methods.
Currently, only API token and secret-key authentication are supported.

## How
- Added a new authenticator `CustomOauth2PrivateKeyAuthenticator` to create the JWT payload and sign it.
- Edited `spec.yaml` and `manifest.yaml` to support the additional schemas.
- Added a sample `config_private_key.json`.

## User Impact
None - No breaking changes.

## Can this PR be safely reverted and rolled back?
- [x] YES 💚
- [ ] NO ❌

### CODE REVIEW

**Comment by natikgadzhi** [comment]
Thank you for this, @evyros! <3 

We'll run some regression tests and get back to you shortly.

File: airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py [hunk 1]
```diff
@@ -2,9 +2,12 @@
 # Copyright (c) 2024 Airbyte, Inc., all rights reserved.
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py:2]: Can you rename the file to `components.py`? It's not validated by any CI checks, but we'll consolidate custom components into that file for all connectors.

File: airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py [hunk 2]
```diff
@@ -61,3 +64,55 @@ def refresh_access_token(self) -> Tuple[str, int]:
...
+    config: Config
+
+    @property
+    def auth_header(self) -> str:
+        return "Authorization"
+
+    @property
+    def token(self) -> str:
+        domain = self.config["domain"]
+        client_id = self.config["credentials"]["client_id"]
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py:84]: I don't know exactly how, but maybe there's a nice way to provide graceful errors if credentials are not provided for any reason? Or will it just fails the `spec` call? /cc @artem1205 @darynaishchenko

File: airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py [hunk 3]
```diff
@@ -61,3 +64,55 @@ def refresh_access_token(self) -> Tuple[str, int]:
...
+                "scope": scope,
+            },
+            headers={"Content-Type": "application/x-www-form-urlencoded"},
+        )
+
+        try:
+            response = token_response.json()
+            return f"Bearer {response['access_token']}"
+        except Exception as e:
+            raise Exception(f"Error while getting access token: {e}") from e
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py:118]: @darynaishchenko is there a specific error class we prefer to a generic Exception?

File: airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml [hunk 4]
```diff
@@ -12,6 +12,10 @@ definitions:
     token_refresh_endpoint: "https://{{ config['domain'] }}.okta.com/oauth2/v1/token"
     grant_type: refresh_token
 
+  custom_oauth_private_key_authenticator:
+    type: CustomAuthenticator
+    class_name: source_okta.custom_authenticators.CustomOauth2PrivateKeyAuthenticator
```
Comment (natikgadzhi) [inline airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml:17]: When you rename the file, make sure to change the reference here as well.

**Review by natikgadzhi — APPROVED** [review]
Overall, I like the change, but I need to verify that the current users are fine.

**Comment by natikgadzhi** [comment]
@clnoll seems to be blocked with "not a git repository" regression test bug

File: airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml (hunk 4 above)
Comment (evyros) [reply airbyte-integrations/connectors/source-okta/source_okta/manifest.yaml]: Done - renamed in `manifest.yaml` and test files.

File: airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py (hunk 1 above)
Comment (evyros) [reply airbyte-integrations/connectors/source-okta/source_okta/custom_authenticators.py:2]: Done

**Comment by clnoll** [comment]
@natikgadzhi that's because the token used by the fork doesn't have correct permissions for getting the `connection-retriever`. Are regression tests required for this connector?

**Comment by evycon** [comment]
> @natikgadzhi that's because the token used by the fork doesn't have correct permissions for getting the `connection-retriever`. Are regression tests required for this connector?
I tested on my local machine to make sure that API token authentication is still working properly.

**Comment by evyros** [comment]
> @natikgadzhi that's because the token used by the fork doesn't have correct permissions for getting the `connection-retriever`. Are regression tests required for this connector?

I tested on my local machine to make sure that API token authentication is still working properly.

**Review by natikgadzhi — APPROVED** [review]

**Pull request merged**